### PR TITLE
[PDI-18440][PDI-18442] Fixes issues with null v.s. empty values

### PR DESCRIPTION
**Pull request by pentaho-lmartins**
This is the proposed fix using three different Kettle properties to override the behavior of `KETTLE_EMPTY_STRING_DIFFERS_FROM_NULL` which is producing different results from previous versions considering version **4.4.2** as the correct behavior.

When `KETTLE_EMPTY_STRING_DIFFERS_FROM_NULL` is set to "Y", three other flags may need to be set to "Y" to accommodate changes in behavior between versions.

- `KETTLE_DO_NOT_NORMALIZE_NULL_STRING_TO_EMPTY` - prevents the **Data grid** step from producing empty values when the "Set empty value" option is not set for a given field
- `KETTLE_XML_EMPTY_TAG_YIELDS_EMPTY_VALUE` - allows the **Filter rows** step to filter rows on an empty values, if not set, the empty value will be converted to a `null` and the comparison between empty and null will yield false, thus not filtering the wanted rows;
- `KETTLE_XML_MISSING_TAG_YIELDS_NULL_VALUE` - causes the **Get XML data** step to yield null values on missing elements and empty values on empty elements, otherwise it will always produce empty values, even when nodes are missing.

Please review but **do not merge**.

@ssamora @ppatricio @pamval

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 35m 17s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine,plugins/xml/core
```
---
:ok_hand: All tests passed!
 
Tests run: **4745**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7232/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 35m 18s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine,plugins/xml/core
```
---
:ok_hand: All tests passed!
 
Tests run: **4748**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7232/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 24m 20s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine,plugins/xml/core
```
---
:ok_hand: All tests passed!
 
Tests run: **4748**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7232/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 22m 35s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine,plugins/xml/core
```
---
:ok_hand: All tests passed!
 
Tests run: **4748**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7232/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: core/src/main/java/org/pentaho/di/core/row/value/ValueMetaBase.java [hunk 1]
```diff
@@ -4033,8 +4033,13 @@ public Object convertDataFromString( String pol, ValueMetaInterface convertMeta,
     Boolean isEmptyAndNullDiffer = convertStringToBoolean(
         Const.NVL( System.getProperty( Const.KETTLE_EMPTY_STRING_DIFFERS_FROM_NULL, "N" ), "N" ) );
 
-    if ( pol == null && isStringValue && isEmptyAndNullDiffer ) {
-      pol = Const.NULL_STRING;
+    Boolean normalizeNullStringToEmpty = !convertStringToBoolean(
+      Const.NVL( System.getProperty( Const.KETTLE_DO_NOT_NORMALIZE_NULL_STRING_TO_EMPTY, "N" ), "N" ) );
+
+    if ( normalizeNullStringToEmpty ) {
+      if ( pol == null && isStringValue && isEmptyAndNullDiffer ) {
```
Comment (ssamora) [inline core/src/main/java/org/pentaho/di/core/row/value/ValueMetaBase.java]: why not if ( normalizeNullStringToEmpty && pol == null && isStringValue && isEmptyAndNullDiffer ), basically just "merging" the two ifs.

Comment (pentaho-lmartins) [reply core/src/main/java/org/pentaho/di/core/row/value/ValueMetaBase.java]: Just for the sake of clarity, so anyone will be able to tell what this change was all about in the not so distant future.

**Pull request merged**